fake-ipa/fake_ipa/main.py

```diff
@@ -0,0 +1,300 @@
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import argparse
+import logging
+import sys
+from threading import Thread
+
+from flask import Flask
+from flask import json
+from flask import request
+from werkzeug.exceptions import HTTPException
+from werkzeug.exceptions import Unauthorized
+from werkzeug import Response
+
+
+from fake_ipa import encoding
+from fake_ipa.fake_agent import FakeIronicPythonAgent
+from fake_ipa.heartbeater import Heatbeater
+
+
+class Application(Flask):
+    agents = {}
+    booted_q = set()
+
+
+app = Application(__name__)
+app.logger.setLevel(logging.DEBUG)
+
+@app.errorhandler(HTTPException)
+def handle_exception(e):
+    """Return JSON instead of HTML for HTTP errors."""
+
+    # start with the correct headers and status code from the error
+    response = e.get_response()
+    # replace the body with JSON
+    response.data = json.dumps({
+        "code": e.code,
+        "name": e.name,
+        "description": e.description,
+    })
+    response.content_type = "application/json"
+    return response
+
+
+@app.route('/', methods=['PUT'])
+def notification_handler():
+    """
+    Endpoint to receive notifications about Systems power state updates
+
+    This function receives a JSON representing a system's current state
+    It boot or destroy FakeIPA for a sytem based on the pending power state
+    recieved
+
+    Example System JSON:
+    {
+        "uuid": "27946b59-9e44-4fa7-8e91-f3527a1ef094",
+        "name": "fake1",
+        "power_state": "On",
+        ...
+        "boot_device": "Pxe",
+        "pending_power": {
+            "power_state": "Off",
+            "apply_time": 1720787353
+        }
+    }
+
+    Processing Logic:
+    1. If the 'pending_power' field is missing or empty, no action is taken.
+    2. If 'pending_power.power_state' is 'On':
+        a. If the system is not already booted, it will be added to the boot queue
+           and the boot process will start after the delay specified by 'apply_time'.
+    3. If 'pending_power.power_state' is not 'On':
+        a. If the system is currently booted, it will be removed from the boot queue
+           and shutdown IPA after the delay specified by 'apply_time'.
+
+    Note:
+    - If the system is already powered on or off, duplicate 'pending_power' state updates
+      should be ignored.
+    """
+    system = request.json
+    system_name = system.get('name', 'unknown')
+    app.logger.info("Received system update for %s: %s", system_name, system)
+
+    # Check if 'pending_power' is present and not None or empty
+    if 'pending_power' not in system or not system['pending_power']:
```

Comment:
Where is this `pending_power` set?

Reply:
coming from sushy-tools notifications

Reply:
Could you point me to the specific location? I cannot find it anywhere, sushy-tools checks this, sure, but it never sets it.
In my trial, this `if` case prevents fake-ipa to ever response to any boot request, because `pending_power` is never set in the request.

Reply:
Regarding power state you are not supposed to send it directly to fakeIPA but only sushy-tools notify it with some here where it is set in the sushy-tools: https://opendev.org/openstack/sushy-tools/src/commit/c5b64a27e1e7ea1a2e5b7490633353f5649ad297/sushy_tools/emulator/resources/systems/fakedriver.py#L142